### PR TITLE
[FLINK-6888] [table] Can not determine TypeInformation of ACC type of AggregateFunction when ACC is a Scala class

### PR DESCRIPTION
**Pull request by wuchong**
Currently the `ACC TypeInformation of `org.apache.flink.table.functions.AggregateFunction[T, ACC]` is extracted using `TypeInformation.of(Class)`. When `ACC` is a Scala case class or tuple class, the TypeInformation will fall back to `GenericType` which result in bad performance when state de/serialization. 

This PR fix this issue by extracting ACC TypeInformation when calling `TableEnvironment.registerFunction()`.

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 1]
```diff
@@ -1395,50 +1399,25 @@ object AggregateUtil {
...
-                throw new TableException(
-                  "Cannot infer type of accumulator. " +
-                    "You can override AggregateFunction.getAccumulatorType() to specify the type.",
-                  ite)
-            }
-          }
-      }
+    // create accumulator type information for every aggregate function
+    aggregates.zipWithIndex.foreach { case (agg, index) =>
+      accTypes(index) = getAccumulatorTypeOfAggregateFunction(agg, accTypes(index))
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: I think we need add a check as follows:
```
aggregates.zipWithIndex.foreach { case (agg, index) =>
      if(null ==  accTypes(index)){
        accTypes(index) = getAccumulatorTypeOfAggregateFunction(agg, accTypes(index))
      }
    }
```

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/AggSqlFunction.scala [hunk 2]
```diff
@@ -37,24 +36,26 @@ import org.apache.flink.table.functions.utils.UserDefinedFunctionUtils._
...
 class AggSqlFunction(
     name: String,
     aggregateFunction: AggregateFunction[_, _],
-    returnType: TypeInformation[_],
+    val returnType: TypeInformation[_],
+    val accType: TypeInformation[_],
     typeFactory: FlinkTypeFactory,
     requiresOver: Boolean)
-  extends SqlUserDefinedAggFunction(
+  extends SqlAggFunction(
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/AggSqlFunction.scala]: Sorry, I didn't notice the copied `SqlUserDefinedAggFunction` from Calcite. I did these change because the constructor of Calcite `SqlUserDefinedAggFunction` doesn't match the calling with additional `requiresOver`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala (hunk 1 above)
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: I don't think so. The `getAccumulatorTypeOfAggregateFunction` can handle the nullable `accTypes(index)`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala [hunk 3]
```diff
@@ -329,6 +337,41 @@ object UserDefinedFunctionUtils {
     }
   }
 
+
+  /**
+    * Internal method of AggregateFunction#getAccumulatorType() that does some pre-checking
+    * and uses [[TypeExtractor]] as default return type inference.
+    */
+  def getAccumulatorTypeOfAggregateFunction(
+    aggregateFunction: AggregateFunction[_, _],
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala]: Good idea. I like it.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala (hunk 1 above)
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: If i understand correctly. only build-in AGG need call `getAccumulatorTypeOfAggregateFunction`. 
and all the UDAGG do not need call this method. is that correct? If so, I think if only `null==accTypes(index)` need call this method. Please explain more If I understand incorrectly. thanks!

Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: Yes, adding the `null == accTypes(index)` condition can reduce one more time calling `getAccumulatorTypeOfAggregateFunction`.  I will add it.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala (hunk 3 above)
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala]: I find that only the following code block can be extracted. I think maybe we don't need to extract it. What do you think? @sunjincheng121 

```
  val resultType = try {
      val method: Method = aggregateFunction.getClass.getMethod(methodName)
      method.invoke(aggregateFunction).asInstanceOf[TypeInformation[_]]
    } catch {
      case _: NoSuchMethodException => null
      case ite: Throwable => throw new TableException("Unexpected exception:", ite)
    }
```

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala [hunk 4]
```diff
@@ -329,6 +337,41 @@ object UserDefinedFunctionUtils {
...
+      case ite: Throwable => throw new TableException("Unexpected exception:", ite)
+    }
+    if (accType != null) {
+      accType
+    } else if (extractedType != null) {
+      extractedType
+    } else {
+      val accumulator = aggregateFunction.createAccumulator()
+      try {
+        TypeInformation.of(accumulator.getClass)
```
Comment (hustfxj) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala]: Maybe we use the followed code replace **TypeInformation.of** ?
 ```
val accumulator = TypeExtractor
        .createTypeInfo(aggregateFunction, classOf[AggregateFunction[_, _]],
          aggregateFunction.getClass, 1).asInstanceOf[TypeInformation[_]]
```

Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala]: I think it's a good advice.

**Comment by StephanEwen** [comment]
Is the code ever called from Java code, or is that purely within the Scala code base?

Just FYI, the design principles in the DataStream and DataSet API are:
  - Never run `TypeExtractor` for Scala code. Scala code always uses implicits and context bounds for the Type Information
  - Java code entry methods uses the Type Extractor
  - Code that is shared between Java / Scala always gets the Type Information passed in

**Comment by wuchong** [comment]
Thanks @StephanEwen .  This PR is aiming to fix the issues you mentioned: it used `TypeExtractor` for Scala code. 

@sunjincheng121  I addressed all your comments. Could you please have a look again ?

**Comment by sunjincheng121** [comment]
@wuchong thanks for the update. It's looks pretty good for me now.
+ to merged.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala [hunk 5]
```diff
@@ -358,20 +358,27 @@ abstract class TableEnvironment(val config: TableConfig) {
...
-    val typeInfo: TypeInformation[_] = implicitly[TypeInformation[T]]
+    val resultTypeInfo: TypeInformation[_] = implicitly[TypeInformation[T]]
+    val accTypeInfo: TypeInformation[_] = implicitly[TypeInformation[ACC]]
 
     // register in Table API
     functionCatalog.registerFunction(name, function.getClass)
 
     // register in SQL API
-    val sqlFunctions = createAggregateSqlFunction(name, function, typeInfo, typeFactory)
+    val sqlFunctions = createAggregateSqlFunction(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala]: To be honest, I don't think it is a nice approach to add the acc type to the internal representation of UDAGGs. It is not related to the logic of the function or required for the optimization but rather an internal runtime aspect.

An alternative could be to wrap the functions registered from Scala in a class that extends the `AggregateFunction` interface and holds the original agg function and the acc type information (exposed via the getAccumulatorType()` method). We would need to unwrap it before we translate it. It's not a super nice solution either, but would probably require fewer changes.

What do you think?

Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala]: Hi @fhueske , I think the ACC type extraction is similar to `T` type of `AggreateFunction[T, ACC]` and `T` type of `TableFunction[T]`. Currently, they are both extracted and add to the internal representation of UDAGGs and UDTFs. To keep the consistent, I followed the same way to add the acc type. If we decide to adopt the approach that wrapping the registered functions, it would be better to make the other type extractions follow this approach.

What do you think ?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/java/BatchTableEnvironment.scala [hunk 6]
```diff
@@ -192,10 +193,14 @@ class BatchTableEnvironment(
       name: String,
       f: AggregateFunction[T, ACC])
   : Unit = {
-    implicit val typeInfo: TypeInformation[T] = TypeExtractor
-      .createTypeInfo(f, classOf[AggregateFunction[T, ACC]], f.getClass, 0)
+    implicit val typeInfo: TypeInformation[T] = UserDefinedFunctionUtils
+      .getResultTypeOfAggregateFunction(f)
       .asInstanceOf[TypeInformation[T]]
 
+    implicit val accTypeInfo: TypeInformation[ACC] = UserDefinedFunctionUtils
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/java/BatchTableEnvironment.scala]: I agree with you. But this PR is only to fix Scala classes can't work in the Scala TableEnvironments. In general, users create Scala classes and naturally will use it in Scala TableEnvironments. Registering a Scala class into Java TableEnvironments is not recommended (should forbidden). This is the same with that Scala `T` classes of `Aggregate[T, _] and TableFunction[T]` works in Scala TableEnvironments.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala (hunk 5 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala]: You are right. We use the same approach for the return type `T` of all UDFs. However, the return type is part of the function signature and needs to be known to Calcite for semantic validation. The `ACC` type is only needed for compilation and an engine specific property.  

TBH, I'm not sure which approach is better.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/java/BatchTableEnvironment.scala (hunk 6 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/java/BatchTableEnvironment.scala]: UDAGGs are sometimes also shared in libraries. If a UDAGG class is included in a JAR file and loaded into the classpath, a user can hardly tell whether it is implemented in Java or Scala. So, I would not prohibit Scala UDFs in Jave environments. However, functions which are intended to be shared should be implemented with Java types.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala (hunk 5 above)
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala]: @fhueske thanks for your explanation. I tried to wrap the UDAGG in an `AggregateFunctionWrapper`, but find that it is not an easy way. Because I can't override the user-defined contract methods, such as `accumulate`, `retract`, `merge`. And in code generation, we generate the acc type class depend on `createAccumulator` method return type, but the return type of `createAccumulator` of `AggregateFunctionWrapper` can only be `Any` which will result in error.

In addition, I plan to support composite result type for UDAGG. This also need the return type not only to be known to Calcite for semantic validation but also to be known for code generation.

Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala]: Ping @fhueske , what do you think about this?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala [hunk 7]
```diff
@@ -320,12 +348,20 @@ object UserDefinedFunctionUtils {
     } else if(extractedType != null) {
       extractedType
     } else {
-      TypeExtractor
+      try {
+        TypeExtractor
```
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala]: I think it is more safe than `TypeInformation.of(Any)`, because `TypeInformation.of(Any)` can only works for non-generic types. 

For example: 

```
public class UDAF extends AggregateFunction[Tuple2<String, Long>, Long] {
    public Long createAccumulator() {...}
    public Tuple2<String, Long> getValue(Long acc) {...}
}
```
For the given UDAF, the return type is a generic type `Tuple2<String, Long>`. We can't extract the return type  from an instance of Tuple2, because of Java type erasure. But we can extract it from the type hierarchy of UDAF class. 

Am I right, @twalthr ?



```

**Comment by wuchong** [comment]
Comments addressed

**Comment by wuchong** [comment]
I have rebased the code and updated.  Would you like to have a look again? @fhueske  @twalthr @sunjincheng121

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala [hunk 8]
```diff
@@ -320,12 +347,20 @@ object UserDefinedFunctionUtils {
     } else if(extractedType != null) {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala]: +space `} else if (extractedType`

File: flink-libraries/flink-table/src/test/java/org/apache/flink/table/api/java/utils/UserDefinedAggFunctions.java [hunk 9]
```diff
@@ -59,7 +59,7 @@ public boolean requiresOver() {
 	/**
 	 * Accumulator for WeightedAvg.
 	 */
-	public static class WeightedAvgAccum extends Tuple2<Long, Integer> {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/table/api/java/utils/UserDefinedAggFunctions.java]: Why do you change the type of this accumulator?

Comment (wuchong) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/table/api/java/utils/UserDefinedAggFunctions.java]: TypeExtractor will recognize `WeightedAvgAccum` as a `TupleType` which will ignore the `sum` and `count` fields when de/serialize. So I changed it as a POJO class. This should not be included in this PR. But I fixed it when I find it.

Comment (fhueske) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/table/api/java/utils/UserDefinedAggFunctions.java]: Yes, you are right. Either we change the type or we touch the code and use `acc.f0` instead of `acc.sum` and `acc.f1` instead of `acc.count`.

Let's go with your approach.

**Comment by fhueske** [comment]
+1 to merge

**Comment by wuchong** [comment]
merging...
